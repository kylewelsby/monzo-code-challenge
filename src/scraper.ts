/**
 * Scrape the website recursively for urls
 *
 * @param inputUrl - the URL to scrape
 * @param maxLevel - the maximum level of recursion
 * @param visitedUrls - visited URLs (used on recursive calls)
 * @param level - the current level of recursion
 * @returns a list of URLs
 */
export async function scrapeRecursive(
  inputUrl: URL,
  maxLevel = 4,
  fileOutput: string | undefined = undefined,
) {
  console.log(`ðŸ”Ž Scraping urls from ${inputUrl.href}`);
  urlQueue.push(inputUrl.href);
  if (fileOutput) {
    /* Clear the file */
    await Deno.writeTextFile(fileOutput, inputUrl.href + "\n", {
      append: false,
    });
  }
  await processQueue(maxLevel, fileOutput);

  console.log(`âœ… Completed with ${visitedUrls.size} urls`);
  return Array.from(visitedUrls);
}

const urlQueue: string[] = [];
const visitedUrls = new Set<string>();

async function processQueue(
  maxLevel = 2,
  fileOutput: string | undefined = undefined,
) {
  let level = 0;
  while (urlQueue.length > 0) {
    const urls = urlQueue.splice(0, Math.min(100, urlQueue.length));
    urls.forEach((url) => visitedUrls.add(url));
    let newUrls;
    if (level < maxLevel) {
      level++;
      newUrls = await recursiveWorker(urls);
      console.log(`ðŸ‘€ Level: ${level} - ${newUrls.length} urls`);
    } else {
      newUrls = urls;
    }
    newUrls.filter((url) => !visitedUrls.has(url) && !urlQueue.includes(url))
      .forEach(async (url) => {
        urlQueue.push(url);
        if (fileOutput) {
          await Deno.writeTextFile(fileOutput, url + "\n", { append: true });
        }
      });

    await new Promise((resolve) => setTimeout(resolve, 100));
  }
}

function recursiveWorker(
  parentUrls: string[],
): Promise<string[]> {
  return new Promise((resolve) => {
    const worker = new Worker(new URL("./worker.ts", import.meta.url).href, {
      type: "module",
    });
    worker.postMessage({
      urls: parentUrls,
    });
    worker.onmessage = (e) => {
      const urls = e.data as string[];
      worker.terminate();
      resolve(urls);
    };
    worker.onerror = (e) => {
      console.error(e);
      worker.terminate();
      resolve([]);
    };
  });
}
