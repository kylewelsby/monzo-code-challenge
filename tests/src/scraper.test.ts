import { scrapeRecursive } from "../../src/scraper.ts";
import { assertEquals, assertSnapshot } from "../../tests/deps.ts";

const MonzoHTML = Deno.readTextFileSync(
  `${Deno.cwd()}/tests/src/__fixtures__/monzo.com.html`,
);

import { serve } from "https://deno.land/std@0.188.0/http/server.ts";

const handler = () => {
  return new Response(MonzoHTML, { status: 200 });
};
serve(handler, { port: 8080 });

Deno.test("scrapeRecursive", async (t) => {
  await t.step("fetches the recursive links for a given url", async () => {
    const urls = await scrapeRecursive(new URL("http://localhost:8080"), 1);
    assertEquals(urls[0], "http://localhost:8080/");
  });

  const urls = await scrapeRecursive(
    new URL("http://localhost:8080"),
    1,
  );
  await assertSnapshot(t, urls);
});
