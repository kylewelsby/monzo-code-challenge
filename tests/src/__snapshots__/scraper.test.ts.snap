export const snapshot = {};

snapshot[`scrapeRecursive 1`] = `
[
  "http://localhost:8080/",
  "http://localhost:8080/accessibility/",
  "http://localhost:8080/features/savings/",
  "http://localhost:8080/features/travel/",
  "http://localhost:8080/flex/",
  "http://localhost:8080/fraud/",
  "http://localhost:8080/i/business",
  "http://localhost:8080/i/security/",
  "http://localhost:8080/loans/",
  "http://localhost:8080/money-worries/",
  "http://localhost:8080/monzo-plus/",
  "http://localhost:8080/monzo-premium/",
  "http://localhost:8080/our-social-programme/",
  "http://localhost:8080/service-quality-results/",
]
`;
